### PR TITLE
[BEACON-415] Support for geofences

### PR DESCRIPTION
**Pull request by williamrijksen**
Ticket: https://enrise.atlassian.net/browse/BEACON-415.

- introduces BABEL
- introduces the support for geofences on iOS (max 20)

Blocked by #13

### CODE REVIEW

File: lib-ng/scanners/base.js [hunk 1]
```diff
@@ -64,26 +58,33 @@ var BaseScanner = function (beaconMapper, beaconRegionMapper, beaconRegionMonito
...
+        }
+        return (version[0] === 5 && version[1] === 0); // Version 5.0.*
+    };
+
     this.handleiOSLocationPermissions = function () {
         // Handle iOS
-        var version = Ti.version.split(".");
         var permissionType = Ti.Geolocation.AUTHORIZATION_ALWAYS;
-        if (isOldTiVersion()) { // Version 5.0.*
+        if (self.isOldTiVersion()) { // Version 5.0.*
```
Comment (adri) [inline lib-ng/scanners/base.js]: how did work with `this` work before? Was this change just in precaution if the function is used somewhere else?

File: lib-ng/scanners/base.js [hunk 2]
```diff
@@ -105,10 +106,30 @@ var BaseScanner = function (beaconMapper, beaconRegionMapper, beaconRegionMonito
...
-        }
+        var realBeacons = knownBeacons.filter(knownBeacon => !knownBeacon.get('is_geofence'));
+        var geofenceBeacons = knownBeacons.filter(knownBeacon => knownBeacon.get('is_geofence'));
+        startScanningOfKnownBeacons(realBeacons);
         self.addAllEventListeners();
+        self.startScanningGeofences(geofenceBeacons);
+    };
+
+    this.startScanningGeofences = function(geofenceBeacons) {
+        // fallback for locations who doesn't have physical-BLE-Beacons
```
Comment (adri) [inline lib-ng/scanners/base.js]: "fallback for locations who **don't** have physical-BLE-Beacons"

File: lib-ng/service/knownBeacons.js [hunk 3]
```diff
@@ -74,15 +74,19 @@ function handleSuccessfullFetchingBeacons(data) {
...
-            beacon_id: beacon.beacon_id
+            beacon_id: beacon.beacon_id,
         });
         if (_.isEmpty(checkBeaconAlreadySaved)) {
             beacon.UUID = beacon.uuid_beacon.toUpperCase();
+            beacon.is_geofence = !_.isUndefined(beacon.is_geofence) ? beacon.is_geofence : false;
             var sensimityKnownBeacon = baseSensimityService.createSensimityModel('KnownBeacon', beacon);
             sensimityKnownBeacon.save();
 
+            if (!sensimityKnownBeacon.get('is_geofence')) {
```
Comment (adri) [inline lib-ng/service/knownBeacons.js]: That means when a "beacon" is a real beacon, we don't fetch the business rules?

File: lib-ng/scanners/geofence.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+var GeofenceLib = require('com.sensimity.ti.geofence'),
+    beaconMapper = require('./../mapper/geofence/beacon'),
+    beaconHandler = require('./../handlers/beaconHandler');
+
+var enteredRegion = (geofenceRegion) => {
+    var beacon = beaconMapper.map(geofenceRegion);
+    beaconHandler.handle(beacon);
+};
+
+const init = () => {
```
Comment (adri) [inline lib-ng/scanners/geofence.js]: Very ES6 :D nice :+1:

File: lib-ng/mapper/geofence/beacon.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+/* jshint ignore:end */
+
+/**
+ * Public functions
+ */
+
+function map(geofenceRegion) {
+    const beaconRaw = geofenceRegion.identifier.split('|');
+
+    return {
```
Comment (adri) [inline lib-ng/mapper/geofence/beacon.js]: The `beacon.get('beacon_id')`is not important here right?

File: lib-ng/scanners/base.js [hunk 6]
```diff
@@ -105,10 +106,30 @@ var BaseScanner = function (beaconMapper, beaconRegionMapper, beaconRegionMonito
...
+        const geofenceLib = require('./../scanners/geofence');
+        geofenceLib.init();
+        geofenceLib.stopMonitoring();
+        geofenceBeacons.forEach((beacon) => {
+            const identifier = `${beacon.get('beacon_id')}|${beacon.get('UUID')}|${beacon.get('major')}|${beacon.get('minor')}`;
+            geofenceLib.startMonitoring({
+                identifier,
+                latitude: beacon.get('latitude'),
+                longitude: beacon.get('longitude'),
+                radius: 100,
```
Comment (adri) [inline lib-ng/scanners/base.js]: Radius 100 is something that works in all cases? Does this need to be configurable?

**Comment by adri** [comment]
I didn't find a restriction on max 20 geofences. Is this something from the OS?

**Comment by adri** [comment]
Verified in the TestbeaconApp. Thanks @williamrijksen.

- Bluetooth has be be enabled for it to work
- Beacons have to be configured using "geofence" in the Sensimity Dashboard

**Comment by adri** [comment]
> I didn't find a restriction on max 20 geofences. Is this something from the OS? 
Indeed a restriction of the OS.

File: lib-ng/scanners/base.js (hunk 1 above)
Comment (williamrijksen) [inline lib-ng/scanners/base.js]: No `this` didn't work, so I changed to the `self`.

File: lib-ng/scanners/base.js (hunk 2 above)
Comment (williamrijksen) [inline lib-ng/scanners/base.js]: Thanks for catching

File: lib-ng/mapper/geofence/beacon.js (hunk 5 above)
Comment (williamrijksen) [inline lib-ng/mapper/geofence/beacon.js]: No indeed, when you check the other mappers, it also doesn't contain a `beacon_id` ;)

File: lib-ng/scanners/base.js (hunk 6 above)
Comment (williamrijksen) [inline lib-ng/scanners/base.js]: Is it an option to make it configurable in the Sensimity Dashboard? Maybe we can create a new ticket in JIRA for it?

**Comment by williamrijksen** [comment]
> Verified in the TestbeaconApp. Thanks @williamrijksen.
> 
> Bluetooth has be be enabled for it to work
> Beacons have to be configured using "geofence" in the Sensimity Dashboard

Indeed for now the BLE should be enabled, but we have to change this strategy by the introduction of geofences. The other notice is answered by https://github.com/Sensimity/com.sensimity.ti.client/pull/11#discussion_r58428336.

**Pull request merged**